SOT-130: Investigate Non-native field arithmetic (#50)
sota-zk-labs/zkp-documents: docs/non_native_field_arithmetic.md

Reviewer: fix all the pmod

# Non-Native Field Arithmetic

**References:**

- [Aztec Emulated Field and Group Operations for Efficient Large Number Arithmetic](https://hackmd.io/@kullervo/cryptographyExplorations/%2F%40kullervo%2FbignumberCircuits)
- [Non-Native Field Arithmetic](https://hackmd.io/@JkY-zACaSqerTtn_UwFjKg/SJZw6x75o)
- [Field selection for recursive SNARKs](https://medium.com/delendum/field-selection-for-recursive-snarks-726ad56c3a3c)
- [Foreign field arithmetic](https://www.youtube.com/watch?v=IBeo5epQoYo)

## Motivation

> Sometimes, we want to represent modulo arithmetic with prime modulus $p$ over our prime finite field $\mathbb{F}_n$ . It's called

Reviewer 2: does "It's" stand for $F_n$ ?

non-native field arithmetic (or some other names like *foreign field, mismatched field, non-aligned field, wrong field*). Below are a
few use cases:

- **For ZK-Rollup after EIP-4844:** After [EIP-4844](eip_4844.md), this operation is required for proof systems that do not
  use [BLS12-381](../terms/bls12-381.md) modulus to deal with blob transaction.
- **Verify cryptographic constructs:** at the protocol level, it is common to verify cryptographic constructs (hashes, signatures,
  proofs) that may be defined over an arbitrary field ($\mathbb{F}_2, \mathbb{F}_q, ...$) inside our field $\mathbb{F}_n$.

> - **Proof aggregation:** merge multiple proofs and create a single proof from them to save resource.

Reviewer 2: the resource

Reviewer: ```suggestion
- **Proof aggregation:** merge multiple proofs and create a single proof from them to save resources.
```

https://dictionary.cambridge.org/dictionary/english/resource

{1}^{\prime})\cdot2^{2B}+ \\
& \quad \ (a_{3}\cdot b_{0}+a_{0}\cdot b_{3}+a_{1}\cdot b_{2}+a_{2}\cdot b_{1}+q_{3}\cdot p_{0}^{\prime}+q_{0}\cdot p_{3}^{\prime}+q_
{1}\cdot p_{2}^{\prime}+q_{2}\cdot p_{1}^{\prime})\cdot2^{3B} \\
& =t_{0}\cdot2^{0B}+t_{1}\cdot2^{1B}+t_{2}\cdot2^{2B}+t_{3}\cdot2^{3B}
\end{aligned}
$$

We can ignore $a_i \cdot b_j$ with $i+j \geq 4$ because it is divisible by $2^T = 2^{4B}$. Now we have:

$$
\begin{array}{l}
{{t_{0}\in\lbrack 0,2^{2B+1})\mathrm{}}}\\
{{t_{1}\in\lbrack 0,2^{2B+2})\mathrm{}}}\\
{{t_{2}\in\lbrack 0,2^{2B+3})\mathrm{}}}\\
{{t_{3}\in\lbrack 0,2^{2B+3})\mathrm{}}}\end{array}
$$

Let $t = t_{0}\cdot2^{0B}+t_{1}\cdot2^{1B}+t_{2}\cdot2^{2B}+t_{3}\cdot2^{3B}$. Note that $t$ itself does not need to be less than
$2^T$. Thus, if $a \cdot b + p' \cdot q = r \mod 2^T$, it is not guaranteed that $t = r$. However, the last $T$ bits of $t$ are
certainly going to equal $r$ (the reason why we choose $2^T$ as our modulus).

![non_native_field_arithmetic](attachments/non_native_field_arithmetic.png)

With $r = \lbrack r_3, r_2, r_1, r_0\rbrack$, we have $r_0$ is the last $B$ bits of $y_0 = t_0 + 0$, i.e $y_0-r_0 = z_0 \cdot 2^B$ for
some

> non-negative $z_0$. Similarly, $r_1$ is the $B$ bits of $y_1 = t_1 + z_0$ with $y_1 - r_1 = z_1 \cdot 2^B$. We can write the all the

Author: I don't understand this, can you add a suggestion ?

> non-negative $z_0$. Similarly, $r_1$ is the $B$ bits of $y_1 = t_1 + z_0$ with $y_1 - r_1 = z_1 \cdot 2^B$. We can write the all the

Author: my bad

limbs of $r$ in a similar fashion:

$$
\begin{array}{l l}
{{y_{0}=t_{0}+0,}}&{{y_{0}-r_{0}=z_{0}\cdot2^{B}}}\\ 
{{y_{1}=t_{1}+z_{0},}}&{{y_{1}-r_{1}=z_{1}\cdot2^{B}}}\\ 
{{y_{2}=t_{2}+z_{1},}}&{{y_{2}-r_{2}=z_{2}\cdot2^{B}}}\\ 
{{y_{2}=t_{3}+z_{2},}}&{{y_{3}-r_{3}=z_{3}\cdot2^{B}}}
\end{array}
$$

Removing the intermediate variables $y_i$, we get:

$$
\begin{array}{l}
{{t_{0}+0-r_{0}=z_{0}\cdot2^{B}}}\\ 
{{t_{1}+z_{0}-r_{1}=z_{1}\cdot2^{B}}}\\ 
{{t_{2}+z_{1}-r_{2}=z_{2}\cdot2^{B}}}\\
{{t_{3}+z_{2}-r_{3}=z_{3}\cdot2^{B}}}
\end{array}
$$

To summarize, for a given $(\lbrack a_3,a_2,a_1,a_0\rbrack, \lbrack b_3, b_2, b_1, b_0\rbrack, \lbrack p'_3, p'_2, p'_1, p'_0\rbrack,
\lbrack q_3, q_2, q_1, q_0\rbrack, \lbrack r_3,r_2,r_1,r_0\rbrack,
t_3,t_2,t_1,t_0, z_3,z_2,z_1,z_0)$ that satisfies the 4 constraints above, along with constraints that ensures $p'=2^T-p$ (note that
$p$ is an odd prime),

> 

Reviewer 2: need linking word here

Author: I don't understand this, can you add a suggestion ?

Reviewer 2: add "and"

$$
\begin{array}{l}
{{p_{0}^{\prime}+p_{0}+0=2^{B}}}\\ 
{{p_{1}^{\prime}+p_{1}+1=2^{B}}}\\ 
{{p_{2}^{\prime}+p_{2}+1=2^{B}}}\\ 
{{p_{3}^{\prime}+p_{3}+1=2^{B}}}
\end{array}
$$

it is guaranteed that $a \cdot b - q \cdot p - r = 0 \mod 2^T$ (but not $a \cdot b - q \cdot p - r = 0$ on their own).

### Constraints in Our Native Field

Next, we will consider a method to ensure $a \cdot b - q \cdot p - r = 0 \mod n$. We denote some new variables $\lbrace a_n, b_n, q_n,
p_n,
r_n\rbrace \ s.t. a_n = a \mod n$ etc. respectively. Now, we have more constraints with advice variables $\lbrace v_a, v_b, v_q, v_p,
v_r, v_
{overall}\rbrace$

$$
\begin{array}
{c}{{v_{a}\cdot n+a_{n}=a}}\\ 

> {{v_{b}\cdot n+b_{n}=a}}\\ 

Reviewer 2: a ?

{{v_{q}\cdot n+q_{n}=p}}\\ 
{{v_{r}\cdot n+r_{n}=r}} \\
v_{overall} \cdot n = a_n \cdot b_n - q_n \cdot p_n - r_n

> \end{array}

Reviewer 2: where is $v_p$ ?

$$

Because $a,b,q,p,r$ are not fit in our native field, we need to decompose them into their limbs. 
For a given $(a,b,q,p,r,v_a, v_b, v_q, v_p, v_r, v_{overall})$ that follow the 5 constraints above, it is guaranteed that $a \cdot b -
q \cdot p - r = 0 \mod n$, but not $a \cdot b - q \cdot p - r = 0$.

### Applying the Chinese Remainder Theorem

Now we have $a \cdot b - q \cdot p - r = 0 \mod n$ and $a \cdot b - q \cdot p - r = 0 \mod 2^T$, which implies that $a \cdot b - q

> \cdot p - r = 0 \mod (n \cdot 2^T)$ by our CRT. If we choose $T$ such that $a \cdot b - q \cdot p - r < p^2 < n \cdot 2^T$, then it

Reviewer 3: ```suggestion
\cdot p - r = 0 \mod (n \cdot 2^T)$ by CRT. If we choose $T$ such that $a \cdot b - q \cdot p - r < p^2 < n \cdot 2^T$, then it
```

means that $a \cdot b - q \cdot p - r = 0$. Note that we should choose $T$ such that $t_3 < n$ to be able to represent all our numbers
in our native field $\mathbb{F}_n$.

## Non-native Field Arithmetic with Lookup Argument

> [!Warning]
> This section is written according to my personal understanding, so it should be used for reference rather than belief

> Because of the "Cost of non-aligned field operations have gone down a lot since the introduction of schemes like Plookup." sentence 

Reviewer 2: because..... but ?
Did you check your grammar before pushing ?
Rewrite this paragraph.
WLS!

# Non-Native Field Arithmetic

**References:**

- [Aztec Emulated Field and Group Operations for Efficient Large Number Arithmetic](https://hackmd.io/@kullervo/cryptographyExplorations/%2F%40kullervo%2FbignumberCircuits)
- [Non-Native Field Arithmetic](https://hackmd.io/@JkY-zACaSqerTtn_UwFjKg/SJZw6x75o)
- [Field selection for recursive SNARKs](https://medium.com/delendum/field-selection-for-recursive-snarks-726ad56c3a3c)
- [Foreign field arithmetic](https://www.youtube.com/watch?v=IBeo5epQoYo)

## Motivation

Sometimes, we want to represent modulo arithmetic with prime modulus $p$ over our prime finite field $\mathbb{F}_n$ . It's called
non-native field arithmetic (or some other names like *foreign field, mismatched field, non-aligned field, wrong field*). Below are a
few use cases:

- **For ZK-Rollup after EIP-4844:** After [EIP-4844](eip_4844.md), this operation is required for proof systems that do not

>   use [BLS12-381](../terms/bls12-381.md) modulus to deal with blob transaction.

Reviewer: ```suggestion
  use [BLS12-381](../terms/bls12-381.md) modulus to deal with blob transactions.
```

Author: "blob transaction" is a transaction type, not a count, so it has no "s".

- **Verify cryptographic constructs:** at the protocol level, it is common to verify cryptographic constructs (hashes, signatures,
  proofs) that may be defined over an arbitrary field ($\mathbb{F}_2, \mathbb{F}_q, ...$) inside our field $\mathbb{F}_n$.
- **Proof aggregation:** merge multiple proofs and create a single proof from them to save resource.

In this article, we will take advantage of [Chinese Remainder Theorem](../terms/chinese_remainder_theorem.md) (CRT)
and [Plookup](reed_solomon_fingerprinting.md) to solve the problem, separately.

## Problem

We are interested in building a circuit for $a \cdot b \pmod p$ over our field $\mathbb{F}_n$ (the same for $a+b$ with our trick).
Typically, a witness $(a,b,r,q,p)$ is produced and the integer constraint $a \cdot b - p \cdot q - r = 0$ is included as part of
the [polynomial commitment](../terms/polynomial-commitment/000_polynomial_commitment.md).

> If we are willing to deal with values in the range $\lbrack 0, p^2)$ in our circuit constraints, we can simply ensure that $a \cdot b -

Reviewer: ```suggestion
We need to avoid the computation of $a \cdot b$ or $p \cdot
 q$, because it is inefficient to compute them if $p$ is a large prime number
```

We need to avoid the computation of $a \cdot b$ or $p \cdot q$, because it is inefficient to compute them if $p$ is a large prime
number.

For simplicity, we can assume that $n < p$.

## Non-native Field Arithmetic with CRT

### Constraint Decomposition with Limbs

> In the following, we will work in modulus $2^T$, where $p < 2^T$. We denote $p' = -p \mod 2^T$, or more concrete $p' = 2^T - p$. It is

Reviewer: pmod

0\rbrack \cdot \lbrack q_3, q_2, q_1, q_0\rbrack \\
& =(a_{0}\cdot b_{0}+q_{0}\cdot p_{0}^{\prime})\cdot2^{0B}+ \\
& \quad \ (a_{1}\cdot b_{0}+a_{0}\cdot b_{1}+q_{1}\cdot p_{0}^{\prime}+q_{0}\cdot p_{1}^{\prime})\cdot2^{1B}+ \\
& \quad \ (a_{2}\cdot b_{0}+a_{0}\cdot b_{2}+a_{1}\cdot b_{1}+q_{2}\cdot p_{0}^{\prime}+q_{0}\cdot p_{2}^{\prime}+q_{1}\cdot p_
{1}^{\prime})\cdot2^{2B}+ \\
& \quad \ (a_{3}\cdot b_{0}+a_{0}\cdot b_{3}+a_{1}\cdot b_{2}+a_{2}\cdot b_{1}+q_{3}\cdot p_{0}^{\prime}+q_{0}\cdot p_{3}^{\prime}+q_
{1}\cdot p_{2}^{\prime}+q_{2}\cdot p_{1}^{\prime})\cdot2^{3B} \\
& =t_{0}\cdot2^{0B}+t_{1}\cdot2^{1B}+t_{2}\cdot2^{2B}+t_{3}\cdot2^{3B}
\end{aligned}
$$

We can ignore $a_i \cdot b_j$ with $i+j \geq 4$ because it is divisible by $2^T = 2^{4B}$. Now we have:

$$
\begin{array}{l}
{{t_{0}\in\lbrack 0,2^{2B+1})\mathrm{}}}\\
{{t_{1}\in\lbrack 0,2^{2B+2})\mathrm{}}}\\
{{t_{2}\in\lbrack 0,2^{2B+3})\mathrm{}}}\\
{{t_{3}\in\lbrack 0,2^{2B+3})\mathrm{}}}\end{array}
$$

Let $t = t_{0}\cdot2^{0B}+t_{1}\cdot2^{1B}+t_{2}\cdot2^{2B}+t_{3}\cdot2^{3B}$. Note that $t$ itself does not need to be less than

> $2^T$. Thus, if $a \cdot b + p' \cdot q = r \mod 2^T$, it is not guaranteed that $t = r$. However, the last $T$ bits of $t$ are

Reviewer: pmod

![non_native_field_arithmetic](attachments/non_native_field_arithmetic.png)

With $r = \lbrack r_3, r_2, r_1, r_0\rbrack$, we have $r_0$ is the last $B$ bits of $y_0 = t_0 + 0$, i.e $y_0-r_0 = z_0 \cdot 2^B$ for
some
non-negative $z_0$. Similarly, $r_1$ is the last $B$ bits of $y_1 = t_1 + z_0$ with $y_1 - r_1 = z_1 \cdot 2^B$. We can write the all
the
limbs of $r$ in a similar fashion:

$$
\begin{array}{l l}
{{y_{0}=t_{0}+0,}}&{{y_{0}-r_{0}=z_{0}\cdot2^{B}}}\\
{{y_{1}=t_{1}+z_{0},}}&{{y_{1}-r_{1}=z_{1}\cdot2^{B}}}\\
{{y_{2}=t_{2}+z_{1},}}&{{y_{2}-r_{2}=z_{2}\cdot2^{B}}}\\
{{y_{2}=t_{3}+z_{2},}}&{{y_{3}-r_{3}=z_{3}\cdot2^{B}}}
\end{array}
$$

Removing the intermediate variables $y_i$, we get:

$$
\begin{array}{l}
{{t_{0}+0-r_{0}=z_{0}\cdot2^{B}}}\\
{{t_{1}+z_{0}-r_{1}=z_{1}\cdot2^{B}}}\\
{{t_{2}+z_{1}-r_{2}=z_{2}\cdot2^{B}}}\\
{{t_{3}+z_{2}-r_{3}=z_{3}\cdot2^{B}}}
\end{array}
$$

To summarize, for a given $(\lbrack a_3,a_2,a_1,a_0\rbrack, \lbrack b_3, b_2, b_1, b_0\rbrack, \lbrack p'_3, p'_2, p'_1, p'_0\rbrack,
\lbrack q_3, q_2, q_1, q_0\rbrack, \lbrack r_3,r_2,r_1,r_0\rbrack,t_3,t_2,t_1,t_0, z_3,z_2,z_1,z_0)$

> that satisfies the 4 constraints above, along with constraints that ensures $p'=2^T-p$ (notice that $p$ is an odd prime)

Reviewer: ```suggestion
that satisfies the 4 constraints above, along with constraints that ensure $p'=2^T-p$ (notice that $p$ is an odd prime)
```

> t_3,t_2,t_1,t_0, z_3,z_2,z_1,z_0)$ that satisfies the 4 constraints above, along with constraints that ensures $p'=2^T-p$ (notice that

Reviewer: ```suggestion
\lbrackt_3,t_2,t_1,t_0\rbrack, \lbrackz_3,z_2,z_1,z_0\rbrack)$ that satisfies the 4 constraints above, along with constraints that ensures $p'=2^T-p$ (notice that
```

$$
\begin{array}{l}
{{p_{0}^{\prime}+p_{0}+0=2^{B}}}\\
{{p_{1}^{\prime}+p_{1}+1=2^{B}}}\\
{{p_{2}^{\prime}+p_{2}+1=2^{B}}}\\
{{p_{3}^{\prime}+p_{3}+1=2^{B}}}
\end{array}
$$

> it is guaranteed that $a \cdot b - q \cdot p - r = 0 \mod 2^T$ (but not $a \cdot b - q \cdot p - r = 0$ on their own).

Reviewer: pmod

### Constraints in Our Native Field

> Next, we will consider a method to ensure $a \cdot b - q \cdot p - r = 0 \mod n$. We denote some new variables $\lbrace a_n, b_n, q_n,

Reviewer: pmod

p_n,

> r_n\rbrace \ s.t.\ a_n = a \mod n$ etc., respectively. Now, we have more constraints with advice variables $\lbrace v_a, v_b, v_q, v_p,

Reviewer: pmod

> r_n\rbrace \ s.t.\ a_n = a \mod n$ etc. respectively. Now, we have more constraints with advice variables $\lbrace v_a, v_b, v_q, v_p,

Reviewer 2: ``` suggestion
a \mod n, etc.
```

### Applying the Chinese Remainder Theorem

Now we have $a \cdot b - q \cdot p - r = 0 \mod n$ and $a \cdot b - q \cdot p - r = 0 \mod 2^T$, which implies that $a \cdot b - q
\cdot p - r = 0 \mod (n \cdot 2^T)$ by our CRT. If we choose $T$ such that $a \cdot b - q \cdot p - r < p^2 < n \cdot 2^T$, then it
means that $a \cdot b - q \cdot p - r = 0$. Note that we should choose $T$ such that $t_3 < n$ to be able to represent all our numbers
in our native field $\mathbb{F}_n$.

## Non-native Field Arithmetic with Lookup Argument

> [!Warning]
> This section is written according to my personal understanding, so it should be used for reference rather than belief

Investigating the statement "Cost of non-aligned field operations have gone down a lot since the introduction of schemes like Plookup."
from this [article](https://notes.ethereum.org/@dankrad/kzg_commitments_in_proofs), I explored [Plookup](plookup.md) but couldn't find
any articles about integrating Lookup with non-aligned field operations. Perhaps it's too straightforward to see?

After exploring limbs decomposition, I think Plookup can be applied directly to non-native field arithmetic. For example, suppose we

> want to represent $6 \cdot 2$ in $\mathbb{F}_5$, we can decompose $6$ into some limbs, and then construct a table containing the

Reviewer: $6 \cdot 2$ ?
